### PR TITLE
Don't allow uncompressed pks in witness addresses

### PR DESCRIPTION
**Pull request by stevenroose**
Solves https://github.com/rust-bitcoin/rust-bitcoin/issues/427.

### CODE REVIEW

File: src/util/address.rs [hunk 1]
```diff
@@ -73,15 +75,14 @@ impl fmt::Display for Error {
             Error::Bech32(ref e) => write!(f, "bech32: {}", e),
             Error::EmptyBech32Payload => write!(f, "the bech32 payload was empty"),
             Error::InvalidWitnessVersion(v) => write!(f, "invalid witness script version: {}", v),
-            Error::InvalidWitnessProgramLength(l) => write!(
-                f,
-                "the witness program must be between 2 and 40 bytes in length: lengh={}",
-                l
+            Error::InvalidWitnessProgramLength(l) => write!(f,
+                "the witness program must be between 2 and 40 bytes in length: lengh={}", l,
```
Comment (sgeisler) [inline src/util/address.rs:79]: Off topic, but maybe fix this `lengh={}`?

**Review by sgeisler — DISMISSED** [review]
Looks good. Kinda annoying to have to return a `Result`, but it's probably not worth the hassle to re-engineer the whole key API *again* :sweat_smile:.

File: src/util/address.rs (hunk 1 above)
Comment (stevenroose) [reply src/util/address.rs:79]: Oh yeah didn't notice. I can add a commit for that.

And yeah agreed that it's a bit sad it returns an error. I hope the explicit documentation of the error case can make people comfortable enough to just `unwrap` it.

**Comment by stevenroose** [comment]
Updated, fixed the typo and put the format and typo change in a separate commit.

**Review by junderw — APPROVED** [review]
LGTM

What about p2wsh?

We decided to parse the witnessScript and run a quick check to see if any chunks are 65 bytes starting with a 0x04, and if there are any we check if it is a valid secp256k1 point. Then error if it is a valid point.

There is probably near 0% chance of some useful data being 65 bytes and starting with 0x04, so maybe even checking for a valid key is overkill.

Thoughts?

**Review by sgeisler — APPROVED** [review]

**Comment by stevenroose** [comment]
> What about p2wsh?

Hmm, I'm a bit reluctant to parse scripts, to be honest. Perhaps this can be considered in a separate PR? I think this PRs change is quite straightforward, so I'd prefer to just merge it. If we want to do something for p2wsh, I think it might benefit from a discussion on IRC or so.

**Comment by junderw** [comment]
sounds fair enough.

Related: `blockdata::Script.to_v0_p2wsh` method should also check before hashing.

**Comment by yancyribbens** [comment]
I agree with @sgeisler that it's unfortunate to need to return a `Result`.  It would be better if the type system didn't allow a uncompressed key to be passed to `p2wpkh()` imo.  That would also avoid needing to check explicitly for `!pk.compressed`.

**Comment by sgeisler** [comment]
@yancyribbens it's a tricky trade-off. We could have a `PrivateKey` trait with the implementations `CompressedPK` and `MaybeCompressedPK` and implement everything generically around the trait. But the added complexity might not be worth it. And it would be a separate PR anyway imo. So I'd like to see this one merged first and then maybe explore ways to fix the underlying problem.

**Comment by yancyribbens** [comment]
@sgeisler I agree that the current PR solves the issue.  It would be nice to see either using trait bounds or possibly wrapping in a enum?  Either way, I agree it's probably a fair amount of work that might not be worth doing for this PR but possibly could be a followup PR.

**Review by apoelstra — APPROVED** [review]
Yeah, I think this approach is fine.

We'll need to revisit compressedness in a bit anyway when we have Taproot keys, and we can consider API changes then.

**Pull request merged**